src/main/kotlin/org/rust/ide/inspections/fixes/AddTypeArguments.kt

```diff
@@ -0,0 +1,66 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.inspections.fixes
+
+import com.intellij.codeInspection.LocalQuickFixAndIntentionActionOnPsiElement
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import org.rust.ide.inspections.getTypeArgumentsAndDeclaration
+import org.rust.lang.core.parser.RustParserUtil
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.RsElement
+import org.rust.lang.core.psi.ext.RsGenericDeclaration
+import org.rust.lang.core.psi.ext.typeParameters
+import org.rust.openapiext.buildAndRunTemplate
+import org.rust.openapiext.createSmartPointer
+
+class AddTypeArguments(element: RsElement) : LocalQuickFixAndIntentionActionOnPsiElement(element) {
+    override fun getText(): String = "Add missing type arguments"
+    override fun getFamilyName() = text
+
+    override fun invoke(
+        project: Project,
+        file: PsiFile,
+        editor: Editor?,
+        startElement: PsiElement,
+        endElement: PsiElement
+    ) {
+        val element = startElement as? RsElement ?: return
+        val (typeArguments, declaration) = getTypeArgumentsAndDeclaration(element) ?: return
+
+        val requiredParameters = declaration.requiredGenericParameters
+        if (requiredParameters.isEmpty()) return
+
+        if (typeArguments?.typeReferenceList?.size ?: 0 >= requiredParameters.size) return
+
+        val factory = RsPsiFactory(project)
+
+        val argumentLifetimes = typeArguments?.lifetimeList?.map { it.text } ?: listOf()
+        val argumentTypes = typeArguments?.typeReferenceList?.map { it.text } ?: listOf()
+        val argumentAssociatedTypes = typeArguments?.assocTypeBindingList?.map { it.text } ?: listOf()
+
+        val missingTypes = requiredParameters.drop(argumentTypes.size).map { it.name ?: "_" }
+
+        val newArguments = argumentLifetimes + argumentTypes + missingTypes + argumentAssociatedTypes
+        val newArgumentList = factory.createTypeArgumentList(newArguments, typeArguments?.coloncolon != null)
+        val replaced = if (typeArguments != null) {
+            typeArguments.replace(newArgumentList) as RsTypeArgumentList
```

Comment:
In this case, we can drop some additional info in existent code like comments, extra whitespaces, etc.
Can we avoid these changes?

Reply:
I added code that preserves the surrounding whitespace and comments. It's not perfect, because the formatter will reformat the code after it's inserted, but it should preserve most important stuff, mainly comments.

Reply:
I wanted to do that originally, however I don't know how. If I just use `typeArguments?.typeReferenceList?.add(factory.createType("Foo"))`, it doesn't change PSI/resulting text. Is there a way to insert the items without creating a new `RsTypeArgumentList`?

Reply:
Oh, it's expected since `typeArguments?.typeReferenceList` returns `List` collection and `add` method just insert element into it. But this list is not connected to PSI structure so sich insertion doesn't modify PSI.
You need to add children psi objects to `typeArguments` element directly.
As an alternative solution, you can use underlying text abstraction - `Document`. In this case, you don't need to construct PSI yourself, just insert necessary text into document directly. Don't forget to commit your changes via `com.intellij.psi.PsiDocumentManager#commitDocument` before `editor?.buildAndRunTemplate` because you need to get already changed PSI

Reply:
Thank you, it is much simpler this way.